### PR TITLE
Implement Multithreading for Enhanced Performance in Custom Check Processing

### PR DESCRIPTION
**Pull request by rajeshpandey2053**
### Description:

This pull request introduces a multithreading solution to enhance the performance of custom check processing in the codebase. The existing codebase comprises various rules applied to multiple fields. During execution, it became evident that parallel processing is necessary at two levels: the field level and the argument level.

At the field level, a check needs to be performed for multiple fields simultaneously. Meanwhile, at the argument level, a check traverses through multiple arguments within a single field. Therefore, nested multithreading is required to fully improve the overall performance of the project.

By leveraging multithreading at both levels, we aim to parallelize the execution of checks, thereby significantly improving performance and efficiency, especially in scenarios involving a large number of checks or resolving URLs.

### Changes:
- [ ] Introduces new functions `process_argument` in  custom_checker.py and `_process_field` in checker.py to segregate the processes for parallel execution.
- [ ] Refactors the existing loop in the run function to utilize multithreading for parallel execution of custom checks.
- [ ] Enhances code readability and maintainability by separating concerns and encapsulating logic into reusable functions.
- [ ] Adds appropriate documentation and comments to explain the multithreading implementation and its benefits.
- [ ] Additional error handling for multi-threaded modules

### Testing:

Extensive testing has been conducted to ensure the correctness and performance of the multithreading solution.
Integration tests have been performed to validate the code's functionality in various scenarios and edge cases.

### Impact:

This change significantly improves the performance of custom check processing, especially in scenarios involving a large number of checks or resolving URLs.
Improved the efficiency of the execution on average from `~100s `to` ~10s`

### CODE REVIEW

File: pyQuARC/code/checker.py [hunk 1]
```diff
@@ -154,43 +155,70 @@ def _check_dependencies_validity(self, dependencies, field_dict):
...
+                executor.submit(
+                    self._process_field,
+                    func,
+                    check,
+                    rule_id,
+                    metadata_content,
+                    field_dict,
+                    result_dict,
+                    rule_mapping,
+                )
```
Comment (xhagrg) [inline pyQuARC/code/checker.py:221]: Wouldn't running this in parallel lead to missing values? Eg:
`result_dict = {}`

Running this method in parallel will pass `result_dict` to the number of parallel method calls. when updating parallelly, wouldn't `result_dict` be missing some elements? 

Not sure if pass by value takes care of the issue. proper testing (unit and manual testing) is required.

File: pyQuARC/code/custom_checker.py [hunk 2]
```diff
@@ -137,24 +177,27 @@ def run(
...
+                    external_data,
+                    external_relation,
+                    invalid_values,
+                    validity,
+                )
+                future_results.append(future)
+
+            # Retrieve results from futures
+            for future in future_results:
+                invalid_values, validity = future.result()
```
Comment (xhagrg) [inline pyQuARC/code/custom_checker.py:199]: Wouldn't sub-threading be an issue?

Comment (rajeshpandey2053) [reply pyQuARC/code/custom_checker.py]: Did manual testing with a curated [list](https://docs.google.com/spreadsheets/d/1A1ivCqU0KPMv3RZ20pDW7Aga27JSZ4j9yCAP7KMA4sY/edit#gid=0) of concept ids, no issue at all.

File: pyQuARC/code/checker.py (hunk 1 above)
Comment (rajeshpandey2053) [reply pyQuARC/code/checker.py]: Did manual unit and integration tests with a curated [list](https://docs.google.com/spreadsheets/d/1A1ivCqU0KPMv3RZ20pDW7Aga27JSZ4j9yCAP7KMA4sY/edit#gid=0) of concept ids, results obtained from pyquarc with and without using multithreading are exactly same.
